In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
!ls "/content/drive/My Drive/CS439 DS project"

 collisions_data.csv	       'Demo Video.gslides'
 cs439project_cleanup.ipynb     finalized_model.pkl
 cs439project_model.ipynb       Heatmap.html
 cs439project_visualize.ipynb   Motor_Vehicle_Collisions_-_Crashes_20251111.csv
'Demo Video.gdoc'	       'Project Proposal.gdoc'


In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

data = pd.read_csv("/content/drive/My Drive/CS439 DS project/Motor_Vehicle_Collisions_-_Crashes_20251111.csv")

/tmp/ipython-input-465060225.py:6: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("/content/drive/My Drive/CS439 DS project/Motor_Vehicle_Collisions_-_Crashes_20251111.csv")


## <font color='crimson'>**Milestone - Data Understanding**</font>

In [4]:
data.head()

CRASH DATE CRASH TIME   BOROUGH ZIP CODE  LATITUDE  LONGITUDE  \
0  09/11/2021       2:39       NaN      NaN       NaN        NaN   
1  03/26/2022      11:45       NaN      NaN       NaN        NaN   
2  11/01/2023       1:29  BROOKLYN  11230.0  40.62179 -73.970024   
3  06/29/2022       6:55       NaN      NaN       NaN        NaN   
4  09/21/2022      13:21       NaN      NaN       NaN        NaN   

                     LOCATION           ON STREET NAME CROSS STREET NAME  \
0                         NaN    WHITESTONE EXPRESSWAY         20 AVENUE   
1                         NaN  QUEENSBORO BRIDGE UPPER               NaN   
2      (40.62179, -73.970024)            OCEAN PARKWAY          AVENUE K   
3                         NaN       THROGS NECK BRIDGE               NaN   
4                         NaN          BROOKLYN BRIDGE               NaN   

  OFF STREET NAME  ...  CONTRIBUTING FACTOR VEHICLE 2  \
0             NaN  ...                    Unspecified   
1             NaN  ...                            NaN   
2             NaN  ...                    Unspecified   
3             NaN  ...                    Unspecified   
4             NaN  ...                    Unspecified   

   CONTRIBUTING FACTOR VEHICLE 3  CONTRIBUTING FACTOR VEHICLE 4  \
0                            NaN                            NaN   
1                            NaN                            NaN   
2                    Unspecified                            NaN   
3                            NaN                            NaN   
4                            NaN                            NaN   

   CONTRIBUTING FACTOR VEHICLE 5  COLLISION_ID  \
0                            NaN       4455765   
1                            NaN       4513547   
2                            NaN       4675373   
3                            NaN       4541903   
4                            NaN       4566131   

                   VEHICLE TYPE CODE 1  VEHICLE TYPE CODE 2  \
0                                Sedan                Sedan   
1                                Sedan                  NaN   
2                                Moped                Sedan   
3                                Sedan        Pick-up Truck   
4  Station Wagon/Sport Utility Vehicle                  NaN   

   VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4 VEHICLE TYPE CODE 5  
0                  NaN                 NaN                 NaN  
1                  NaN                 NaN                 NaN  
2                Sedan                 NaN                 NaN  
3                  NaN                 NaN                 NaN  
4                  NaN                 NaN                 NaN  

[5 rows x 29 columns]

In [5]:
data.shape

(2219657, 29)

In [6]:
data.columns.to_list()

['CRASH DATE',
 'CRASH TIME',
 'BOROUGH',
 'ZIP CODE',
 'LATITUDE',
 'LONGITUDE',
 'LOCATION',
 'ON STREET NAME',
 'CROSS STREET NAME',
 'OFF STREET NAME',
 'NUMBER OF PERSONS INJURED',
 'NUMBER OF PERSONS KILLED',
 'NUMBER OF PEDESTRIANS INJURED',
 'NUMBER OF PEDESTRIANS KILLED',
 'NUMBER OF CYCLIST INJURED',
 'NUMBER OF CYCLIST KILLED',
 'NUMBER OF MOTORIST INJURED',
 'NUMBER OF MOTORIST KILLED',
 'CONTRIBUTING FACTOR VEHICLE 1',
 'CONTRIBUTING FACTOR VEHICLE 2',
 'CONTRIBUTING FACTOR VEHICLE 3',
 'CONTRIBUTING FACTOR VEHICLE 4',
 'CONTRIBUTING FACTOR VEHICLE 5',
 'COLLISION_ID',
 'VEHICLE TYPE CODE 1',
 'VEHICLE TYPE CODE 2',
 'VEHICLE TYPE CODE 3',
 'VEHICLE TYPE CODE 4',
 'VEHICLE TYPE CODE 5']

In [7]:
data.describe()

LATITUDE     LONGITUDE  NUMBER OF PERSONS INJURED  \
count  1.979311e+06  1.979311e+06               2.219639e+06   
mean   4.059025e+01 -7.368399e+01               3.281957e-01   
std    2.330352e+00  4.331881e+00               7.149202e-01   
min    0.000000e+00 -2.013600e+02               0.000000e+00   
25%    4.066734e+01 -7.397453e+01               0.000000e+00   
50%    4.072032e+01 -7.392674e+01               0.000000e+00   
75%    4.076956e+01 -7.386657e+01               0.000000e+00   
max    4.334444e+01  0.000000e+00               4.300000e+01   

       NUMBER OF PERSONS KILLED  NUMBER OF PEDESTRIANS INJURED  \
count              2.219626e+06                   2.219657e+06   
mean               1.585402e-03                   5.958984e-02   
std                4.206432e-02                   2.500402e-01   
min                0.000000e+00                   0.000000e+00   
25%                0.000000e+00                   0.000000e+00   
50%                0.000000e+00                   0.000000e+00   
75%                0.000000e+00                   0.000000e+00   
max                8.000000e+00                   2.700000e+01   

       NUMBER OF PEDESTRIANS KILLED  NUMBER OF CYCLIST INJURED  \
count                  2.219657e+06               2.219657e+06   
mean                   7.843554e-04               2.918063e-02   
std                    2.863184e-02               1.705832e-01   
min                    0.000000e+00               0.000000e+00   
25%                    0.000000e+00               0.000000e+00   
50%                    0.000000e+00               0.000000e+00   
75%                    0.000000e+00               0.000000e+00   
max                    6.000000e+00               4.000000e+00   

       NUMBER OF CYCLIST KILLED  NUMBER OF MOTORIST INJURED  \
count              2.219657e+06                2.219657e+06   
mean               1.261456e-04                2.347849e-01   
std                1.127079e-02                6.760680e-01   
min                0.000000e+00                0.000000e+00   
25%                0.000000e+00                0.000000e+00   
50%                0.000000e+00                0.000000e+00   
75%                0.000000e+00                0.000000e+00   
max                2.000000e+00                4.300000e+01   

       NUMBER OF MOTORIST KILLED  COLLISION_ID  
count               2.219657e+06  2.219657e+06  
mean                6.451447e-04  3.265184e+06  
std                 2.778085e-02  1.509542e+06  
min                 0.000000e+00  2.200000e+01  
25%                 0.000000e+00  3.191035e+06  
50%                 0.000000e+00  3.746120e+06  
75%                 0.000000e+00  4.301250e+06  
max                 5.000000e+00  4.856467e+06

Checking dataset for missing values

In [8]:
missing_values = data.isnull().sum()

missing_values_percentage = (missing_values / len(data)) * 100

# Return counts and percentages of missing values in each column
missing_data = pd.DataFrame({'Missing Values': missing_values, 'Percentage (%)': missing_values_percentage})
missing_data.sort_values(by='Percentage (%)', ascending=False)

Missing Values  Percentage (%)
VEHICLE TYPE CODE 5                   2209932       99.561869
CONTRIBUTING FACTOR VEHICLE 5         2209616       99.547633
VEHICLE TYPE CODE 4                   2184322       98.408087
CONTRIBUTING FACTOR VEHICLE 4         2182991       98.348123
VEHICLE TYPE CODE 3                   2065294       93.045637
CONTRIBUTING FACTOR VEHICLE 3         2059062       92.764873
OFF STREET NAME                       1828634       82.383630
CROSS STREET NAME                      848140       38.210408
ZIP CODE                               680402       30.653475
BOROUGH                                680127       30.641086
ON STREET NAME                         483397       21.778004
VEHICLE TYPE CODE 2                    445154       20.055081
CONTRIBUTING FACTOR VEHICLE 2          356262       16.050318
LOCATION                               240346       10.828069
LATITUDE                               240346       10.828069
LONGITUDE                              240346       10.828069
VEHICLE TYPE CODE 1                     16146        0.727410
CONTRIBUTING FACTOR VEHICLE 1            7837        0.353073
NUMBER OF PERSONS KILLED                   31        0.001397
NUMBER OF PERSONS INJURED                  18        0.000811
CRASH DATE                                  0        0.000000
NUMBER OF PEDESTRIANS INJURED               0        0.000000
CRASH TIME                                  0        0.000000
NUMBER OF MOTORIST KILLED                   0        0.000000
NUMBER OF MOTORIST INJURED                  0        0.000000
NUMBER OF PEDESTRIANS KILLED                0        0.000000
NUMBER OF CYCLIST INJURED                   0        0.000000
NUMBER OF CYCLIST KILLED                    0        0.000000
COLLISION_ID                                0        0.000000

Upon a brief look at our data, and degree of missingness in the values, we can conclude that:
* Even though **Collision ID**, has 0 missing values, it's purpose is to assign a unique identifier to each collision. Since, it does not impart any knowledge about the collision and is just an lookup identifier, we can drop this feature.

* Since, **zip codes** are related to the boroughs, and also does not impart any extra knowledge about the crash/collision, we can drop this feature too.

* **Cross street names** (38.21% missing values), **On street names** (21.77% missing values), and **Off street names** (82.38% missing values) are highly unique and are just for reference to find the collision spot, which can also be found using latitude, longitude pair. Therefore, we can drop these features.

* **Contributing Factor Vehicle 3, 4, & 5** has 92.76%, 98.34%, and 99.54% missing values. Using these features for data exploration purposes or supervised learning will not impart any extra knowledge to us or our model. Therefore, we will also drop these 3 features.

* Also, **Location** is just a tuple created by latitude and longitude. Therefore, we will remove this feature to remove redundancy in data.

* Also, before dropping any of these features, let's have a brief look at the feature **Contributing Factor Vehicle 2**, if it's entries seems vague and misleading, we will also drop this feature.

In [9]:
data['CONTRIBUTING FACTOR VEHICLE 2'].unique()

array(['Unspecified', nan, 'Driver Inattention/Distraction',
       'Other Vehicular', 'Traffic Control Disregarded',
       'Passing or Lane Usage Improper', 'Reaction to Uninvolved Vehicle',
       'Following Too Closely', 'Unsafe Speed',
       'Aggressive Driving/Road Rage', 'Turning Improperly',
       'Failure to Yield Right-of-Way', 'Driver Inexperience',
       'View Obstructed/Limited', 'Oversized Vehicle',
       'Passing Too Closely', 'Unsafe Lane Changing', 'Pavement Slippery',
       'Pedestrian/Bicyclist/Other Pedestrian Error/Confusion',
       'Fell Asleep', 'Backing Unsafely', 'Obstruction/Debris',
       'Alcohol Involvement', 'Outside Car Distraction',
       'Passenger Distraction', 'Glare', 'Vehicle Vandalism',
       'Traffic Control Device Improper/Non-Working', 'Tinted Windows',
       'Tow Hitch Defective', 'Brakes Defective', 'Steering Failure',
       'Pavement Defective', 'Failure to Keep Right',
       'Driverless/Runaway Vehicle', 'Drugs (illegal)', 'Fatig

In [10]:
# Let's analyze how much percentage of this feature is 'Unspecified
unspecified_count_vehicle_factor_2 = data[data['CONTRIBUTING FACTOR VEHICLE 2']=='Unspecified'].shape[0]
data_count = data.shape[0]
unspecified_percentage = (unspecified_count_vehicle_factor_2/data_count)*100
vehicle_factor_2_missing_percentage = 16.050318
print(f"Percentage of CONTRIBUTING FACTOR VEHICLE 2 which is unsepcified is {unspecified_percentage}")
print(f"Total irrelevant data in CONTRIBUTING FACTOR VEHICLE 2 is {vehicle_factor_2_missing_percentage + unspecified_percentage}")

Percentage of CONTRIBUTING FACTOR VEHICLE 2 which is unsepcified is 70.64010340336367
Total irrelevant data in CONTRIBUTING FACTOR VEHICLE 2 is 86.69042140336367


Therefore, we can conclude that Contributing Factor Vehicle 2 is also **irrelevant** for us.

---

Now, let's drop the irrelevant features.

In [11]:
data = data.drop(columns=[
    'COLLISION_ID',
    'ZIP CODE',
    'ON STREET NAME',
    'CROSS STREET NAME',
    'OFF STREET NAME',
    'LOCATION',
    'CONTRIBUTING FACTOR VEHICLE 2',
    'CONTRIBUTING FACTOR VEHICLE 3',
    'CONTRIBUTING FACTOR VEHICLE 4',
    'CONTRIBUTING FACTOR VEHICLE 5'
    ])

In [12]:
# Let's take a sneak peek at the data again
data.head()

CRASH DATE CRASH TIME   BOROUGH  LATITUDE  LONGITUDE  \
0  09/11/2021       2:39       NaN       NaN        NaN   
1  03/26/2022      11:45       NaN       NaN        NaN   
2  11/01/2023       1:29  BROOKLYN  40.62179 -73.970024   
3  06/29/2022       6:55       NaN       NaN        NaN   
4  09/21/2022      13:21       NaN       NaN        NaN   

   NUMBER OF PERSONS INJURED  NUMBER OF PERSONS KILLED  \
0                        2.0                       0.0   
1                        1.0                       0.0   
2                        1.0                       0.0   
3                        0.0                       0.0   
4                        0.0                       0.0   

   NUMBER OF PEDESTRIANS INJURED  NUMBER OF PEDESTRIANS KILLED  \
0                              0                             0   
1                              0                             0   
2                              0                             0   
3                              0                             0   
4                              0                             0   

   NUMBER OF CYCLIST INJURED  NUMBER OF CYCLIST KILLED  \
0                          0                         0   
1                          0                         0   
2                          0                         0   
3                          0                         0   
4                          0                         0   

   NUMBER OF MOTORIST INJURED  NUMBER OF MOTORIST KILLED  \
0                           2                          0   
1                           1                          0   
2                           1                          0   
3                           0                          0   
4                           0                          0   

  CONTRIBUTING FACTOR VEHICLE 1                  VEHICLE TYPE CODE 1  \
0  Aggressive Driving/Road Rage                                Sedan   
1             Pavement Slippery                                Sedan   
2                   Unspecified                                Moped   
3         Following Too Closely                                Sedan   
4           Passing Too Closely  Station Wagon/Sport Utility Vehicle   

  VEHICLE TYPE CODE 2 VEHICLE TYPE CODE 3 VEHICLE TYPE CODE 4  \
0               Sedan                 NaN                 NaN   
1                 NaN                 NaN                 NaN   
2               Sedan               Sedan                 NaN   
3       Pick-up Truck                 NaN                 NaN   
4                 NaN                 NaN                 NaN   

  VEHICLE TYPE CODE 5  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN

## <font color='crimson'>**Milestone - Data Preparation**</font>

For the purposes of data preparation, Let's:

> * have a look at the following features:
>
>> * Borough
>> * Vehicle Type Code 1
>> * Contributing Factor Vehicle 1
>
> * understand the these features
>
> * modify them accordingly to maintain consistency of data

---
'BOROUGH'

In [13]:
data['BOROUGH'].unique()

array([nan, 'BROOKLYN', 'BRONX', 'MANHATTAN', 'QUEENS', 'STATEN ISLAND'],
      dtype=object)

This is already consistent; No need to modify

---
'VEHICLE TYPE CODE 1'

In [14]:
data['VEHICLE TYPE CODE 1'].nunique()

1832

In [15]:
data['VEHICLE TYPE CODE 1'].value_counts().head(40)

VEHICLE TYPE CODE 1
Sedan                                  641446
Station Wagon/Sport Utility Vehicle    502450
PASSENGER VEHICLE                      416206
SPORT UTILITY / STATION WAGON          180291
Taxi                                    55825
4 dr sedan                              40187
Pick-up Truck                           38212
TAXI                                    31911
Box Truck                               26527
VAN                                     25266
Bus                                     24267
OTHER                                   22968
UNKNOWN                                 19940
Bike                                    18086
LARGE COM VEH(6 OR MORE TIRES)          14397
BUS                                     13993
SMALL COM VEH(4 TIRES)                  13216
PICK-UP TRUCK                           11505
Tractor Truck Diesel                    11319
LIVERY VEHICLE                          10481
Motorcycle                               9849
Van                                      9826
Ambulance                                5178
Dump                                     4261
MOTORCYCLE                               4195
E-Bike                                   4189
Convertible                              4184
Moped                                    3107
PK                                       2882
AMBULANCE                                2762
E-Scooter                                2737
Flat Bed                                 2653
2 dr sedan                               2653
Garbage or Refuse                        2600
Carry All                                2095
Tractor Truck Gasoline                   1693
Tow Truck / Wrecker                      1517
Chassis Cab                               997
Tanker                                    901
FIRE TRUCK                                894
Name: count, dtype: int64

To ensure consistency in data, let's perform the following conversions:
* 4 dr sedan -> sedan

* 2 dr sedan -> sedan

* Station Wagon/Sport Utility Vehicle -> st wagon/suv

* SPORT UTILITY / STATION WAGON -> st wagon/suv

* MOTORCYCLE -> bike

* Motorcycle -> bike

* Garbage or Refuse -> dump

* Tractor Truck Diesel -> tractor truck gasoline

* motorbike -> bike

* ambul -> ambulance

* motorscooter -> moped

* scooter -> moped

* e-bike -> moped

* anything which has 'moped' with moped

* anything which has 'bus' with bus

* tanker -> tractor truck gasoline

* box truck -> small com veh

* small com veh(4 tires) -> small com veh

* large com veh(6 or more tires) -> large com veh

* tractor truck gasoline -> large com veh

* convert all to lowercase

In [16]:
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.lower()
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('4 dr sedan', 'sedan')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('2 dr sedan', 'sedan')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('station wagon/sport utility vehicle', 'st wagon/suv')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('sport utility / station wagon', 'st wagon/suv')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('motorcycle', 'bike')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('garbage or refuse', 'dump')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('tractor truck diesel', 'tractor truck gasoline')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('motorbike', 'bike')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('motorscooter', 'moped')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('scooter', 'moped')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('tanker', 'tractor truck gasoline')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('e-bike', 'moped')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('box truck', 'small com veh')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('small com veh(4 tires)', 'small com veh')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('large com veh(6 or more tires)', 'large com veh')
data['VEHICLE TYPE CODE 1'] = data['VEHICLE TYPE CODE 1'].str.replace('tractor truck gasoline', 'large com veh')

In [17]:
conditions = [
    # Condition 1: Check if 'VEHICLE TYPE CODE 1' contains 'ambul'
    data['VEHICLE TYPE CODE 1'].str.contains('ambul', case=False, na=False),

    # Condition 2: Check if 'VEHICLE TYPE CODE 1' contains 'bus'
    data['VEHICLE TYPE CODE 1'].str.contains('bus', case=False, na=False),

    # Condition 3: Check if 'VEHICLE TYPE CODE 1' contains 'moped'
    data['VEHICLE TYPE CODE 1'].str.contains('moped', case=False, na=False),

    # Condition 4: Check if 'VEHICLE TYPE CODE 1' contains 'ambul'
    data['VEHICLE TYPE CODE 1'].str.contains('ambul', case=False, na=False)
]

choices = [
    'ambulance', # If condition 1 is True
    'bus',        # If condition 2 is True
    'moped',      # If condition 3 is True
    'ambulance'   # If condition 4 is True
]

data['VEHICLE TYPE CODE 1'] = np.select(
    conditions,
    choices,
    default=data['VEHICLE TYPE CODE 1']
)

In [18]:
top_vehicles = data['VEHICLE TYPE CODE 1'].value_counts().head(17).index.to_list()
condition_to_replace = ~data['VEHICLE TYPE CODE 1'].str.lower().isin(top_vehicles)
data.loc[condition_to_replace, 'VEHICLE TYPE CODE 1'] = 'other'
data['VEHICLE TYPE CODE 1'].value_counts()

VEHICLE TYPE CODE 1
sedan                684297
st wagon/suv         682741
passenger vehicle    416206
taxi                  87740
other                 66374
pick-up truck         49717
bus                   38637
van                   35557
bike                  32712
large com veh         28310
small com veh         26576
unknown               19960
small com veh         13216
moped                 11470
livery vehicle        10481
ambulance              8684
dump                   6979
Name: count, dtype: int64

---
'CONTRIBUTING FACTOR VEHICLE 1'

In [19]:
data['CONTRIBUTING FACTOR VEHICLE 1'].unique()

array(['Aggressive Driving/Road Rage', 'Pavement Slippery', 'Unspecified',
       'Following Too Closely', 'Passing Too Closely',
       'Failure to Yield Right-of-Way', nan, 'Driver Inexperience',
       'Passing or Lane Usage Improper', 'Turning Improperly',
       'Unsafe Lane Changing', 'Unsafe Speed',
       'Reaction to Uninvolved Vehicle', 'Steering Failure',
       'Traffic Control Disregarded', 'Other Vehicular',
       'Driver Inattention/Distraction', 'Accelerator Defective',
       'Oversized Vehicle',
       'Pedestrian/Bicyclist/Other Pedestrian Error/Confusion',
       'Alcohol Involvement', 'View Obstructed/Limited', 'Illnes',
       'Lost Consciousness', 'Brakes Defective', 'Backing Unsafely',
       'Glare', 'Passenger Distraction', 'Fell Asleep',
       'Obstruction/Debris', 'Tinted Windows', 'Animals Action',
       'Drugs (illegal)', 'Pavement Defective', 'Other Lighting Defects',
       'Outside Car Distraction', 'Driverless/Runaway Vehicle',
       'Tire Failure/

Upon a brief look at the data above, we have to perform the following modifications:

* '80' -> unspecified

* '1' -> unspecified

* Illnes -> Illness

In [20]:
data['CONTRIBUTING FACTOR VEHICLE 1'] = data['CONTRIBUTING FACTOR VEHICLE 1'].str.lower()

conditions_factor = [
    data['CONTRIBUTING FACTOR VEHICLE 1'].str.contains('80', case=False, na=False),

    data['CONTRIBUTING FACTOR VEHICLE 1'].str.contains('1', case=False, na=False),

    data['CONTRIBUTING FACTOR VEHICLE 1'].str.contains('illnes', case=False, na=False)
]

choices_factor = [
    'unspecified',
    'unspecified',
    'illness'
]

data['CONTRIBUTING FACTOR VEHICLE 1'] = np.select(
    conditions_factor,
    choices_factor,
    default=data['CONTRIBUTING FACTOR VEHICLE 1']
)

In [21]:
data['CONTRIBUTING FACTOR VEHICLE 1'].value_counts().head(10)

CONTRIBUTING FACTOR VEHICLE 1
unspecified                       743131
driver inattention/distraction    450549
failure to yield right-of-way     132938
following too closely             119230
backing unsafely                   81023
other vehicular                    69355
passing or lane usage improper     63699
passing too closely                56378
turning improperly                 54617
fatigued/drowsy                    47556
Name: count, dtype: int64

## <font color='crimson'>**Milestone - Feature Engineering**</font>

For the purposes of Feature Engineering, we need to build new features to calculate:

> * Multi-vehicle crashes:
>
>> * Find number of vehicles involved in each crash
>> * Features involved => 'VEHICLE TYPE CODE 1', 'VEHICLE TYPE CODE 2', 'VEHICLE TYPE CODE 3', 'VEHICLE TYPE CODE 4', 'VEHICLE TYPE CODE 5'
>
> * Find degree of emergency
>
>> * Find how many people are killed or injured
>> * Categorize crash as 0, 1, 2 signifying low, medium and high degree of emergency depending on no injuries/fatalties, no fatalties but injuries, and fatalties respectively.
>> * Features involved => 'NUMBER OF PERSONS INJURED', 'NUMBER OF PERSONS KILLED', 'NUMBER OF PEDESTRIANS INJURED', 'NUMBER OF PEDESTRIANS KILLED', 'NUMBER OF CYCLIST INJURED', 'NUMBER OF CYCLIST KILLED', 'NUMBER OF MOTORIST INJURED', 'NUMBER OF MOTORIST KILLED'
>
> * Finding Crash Hour
>
>> * Features involved => 'CRASH TIME'

---
Number of Vehicles involved

In [22]:
def count_vehicles_involved(data: pd.DataFrame) -> pd.DataFrame:
    """
    Calculates the number of vehicles involved in each crash by counting
    the number of non-missing values in the specified vehicle code columns.
    """

    # 1. Define the relevant vehicle columns
    vehicle_cols = [
        'VEHICLE TYPE CODE 1',
        'VEHICLE TYPE CODE 2',
        'VEHICLE TYPE CODE 3',
        'VEHICLE TYPE CODE 4',
        'VEHICLE TYPE CODE 5'
    ]

    # 2. Select only the vehicle columns
    vehicle_data = data[vehicle_cols]

    # 3. Check for non-missing values (returns True/1 for non-NA, False/0 for NA)
    # The .notna() method is the robust way to check for anything but NA.
    is_involved = vehicle_data.notna()

    # 4. Sum the True (1) values across the rows (axis=1)
    # The sum of booleans (True=1, False=0) gives the total count.
    data['VEHICLES INVOLVED'] = is_involved.sum(axis=1)

    return data


data = count_vehicles_involved(data)

In [23]:
data = data.drop(columns = ['VEHICLE TYPE CODE 2',
                            'VEHICLE TYPE CODE 3',
                            'VEHICLE TYPE CODE 4',
                            'VEHICLE TYPE CODE 5'])

In [24]:
# Confirming whether the new feature has been created or not
data['VEHICLES INVOLVED']

0          2
1          1
2          3
3          2
4          1
          ..
2219652    2
2219653    2
2219654    2
2219655    1
2219656    2
Name: VEHICLES INVOLVED, Length: 2219657, dtype: int64

---
Degree of Emergency

In [25]:
conditions_emergency = [
    data['NUMBER OF PERSONS KILLED'] > 0,

    data['NUMBER OF PERSONS INJURED'] > 0
]

choices_emergency = [
    2,
    1
]

data['DEGREE OF EMERGENCY'] = np.select(
    conditions_emergency,
    choices_emergency,
    default=0
)

In [26]:
data = data.drop(columns=[
    'NUMBER OF PERSONS INJURED',
    'NUMBER OF PERSONS KILLED',
    'NUMBER OF PEDESTRIANS INJURED',
    'NUMBER OF PEDESTRIANS KILLED',
    'NUMBER OF CYCLIST INJURED',
    'NUMBER OF CYCLIST KILLED',
    'NUMBER OF MOTORIST INJURED',
    'NUMBER OF MOTORIST KILLED'
])

In [27]:
# Confirming whether the new feature has been created or not
data['DEGREE OF EMERGENCY']

0          1
1          1
2          1
3          0
4          0
          ..
2219652    0
2219653    0
2219654    0
2219655    1
2219656    1
Name: DEGREE OF EMERGENCY, Length: 2219657, dtype: int64

---
Finding Crash Hour

In [28]:
data['CRASH TIME'] = pd.to_datetime(data['CRASH TIME'], format='%H:%M')

data['CRASH HOUR'] = data['CRASH TIME'].dt.hour

In [29]:
data = data.drop(columns=['CRASH TIME'])

In [30]:
# Confirming whether the new feature has been created or not
data['CRASH HOUR']

0           2
1          11
2           1
3           6
4          13
           ..
2219652    22
2219653     0
2219654    23
2219655    19
2219656    14
Name: CRASH HOUR, Length: 2219657, dtype: int32

---
Rename columns into a meaningful names

In [31]:
data.rename(columns={
    'CONTRIBUTING FACTOR VEHICLE 1': 'CRASH FACTOR',
    'VEHICLE TYPE CODE 1': 'VEHICLE TYPE'
    }, inplace=True)

---
##<font color='blue'> Export the updated dataset to construct dashboard </font>

In [32]:
# data.to_csv('collisions_data.csv')